File: CS284/week9/HW/mergesort.py
# Takes an array and separates it, returns the 2 separate arrays
def separate (array):
	length = len(array)
	middle = length//2
	
	left_array = array[:middle]
	right_array = array[middle:]

	return left_array, right_array
	


# Takes 2 arrays as input, returns 1 merged array
def merge (array1, array2):
	left_size = len(array1)
	right_size = len(array2)
	merged = []
	lp = 0
	rp = 0
	# Compare and append to merge
	while (lp < left_size and rp < right_size):
		if (array1[lp] < array2[rp]):
			merged.append(array1[lp])
			lp+=1
		else:
			merged.append(array2[rp])
			rp+=1
	
	# Check to see which one has remaining elements, and append all of it to merge
	if (lp < left_size):
		merged.extend(array1[lp:])
	if (rp < right_size):
		merged.extend(array2[rp:])

	return merged

# Takes an array, recursively separates it until the length is 1 (base case) and then merge it
def merge_sort (array):
	if (len(array) <= 1):
		return array

	left, right = separate(array)

	left_sorted = merge_sort(left)
	right_sorted = merge_sort(right)

	return merge(left_sorted, right_sorted)

def main ( ):

	data = []
	item = int(input())
	while (item != -1):
		data.append(item)
		item = int(input())
	
	merged_array = merge_sort(data)
	print(merged_array)
	
main()

# Get middle of array, split array into both left and right half continuously until there is 1. And then 
# Merge together with 2 pointers to determine which element goes which way

File: CS284/week9/HW/mergesort.cpp
#include <bits/stdc++.h>
using namespace std;
// Prototypes
vector<int> mergeSort (vector<int> array);
pair<vector<int>, vector<int>> separate (vector<int> array);
vector<int> merge (vector<int> array1, vector<int> array2);

// Merge sort function
vector<int> mergeSort (vector<int> array)
{
	if (array.size() <= 1) return array;

	pair<vector<int>, vector<int>> separatedArrays = separate(array);

	vector<int> leftSorted = mergeSort(separatedArrays.first);
	vector<int> rightSorted = mergeSort(separatedArrays.second);

	return merge(leftSorted, rightSorted);
}




// Function that returns a pair of 2 halves of an array
pair<vector<int>, vector<int>> separate (vector<int> array)
{
	pair<vector<int>, vector<int>> separated;
	int mid = array.size()/2;
	
	separated.first = vector<int>(array.begin(), array.begin() + mid);
	separated.second = vector<int>(array.begin() + mid, array.end());


	return separated;
}


// Merge, takes in 2 vectors and sorts them with 2 ptrs
vector<int> merge (vector<int> array1, vector<int> array2)
{
	int lp = 0;
	int rp = 0;
	int leftSize = array1.size();
	int rightSize = array2.size();
	vector<int> merged;


	// Merge
	while (lp < leftSize && rp < rightSize)
	{
		if (array1[lp] < array2[rp])
		{
			merged.push_back(array1[lp]);
			++lp;
		}
		else
		{
			merged.push_back(array2[rp]);
			++rp;
		}
	}
	
	// Push the rest 
	if (lp < leftSize)
	{
		merged.insert(merged.end(), array1.begin() + lp, array1.end());
	}
	if (rp < rightSize)
	{

		merged.insert(merged.end(), array1.begin() + rp, array1.end());
	}

	return merged;
}



int main ( )
{
	vector<int> array;
	int num = 0;
	do 
	{
		cout << "Enter a number (-1 TO CANCEL): ";
		cin >> num; cin.ignore(100, '\n');
		if (num != -1) array.push_back(num);
	} while (num != -1);

	vector<int> sorted = mergeSort(array);

	cout << "Sorted array: \n";
	for (const auto x : sorted)
	{
		cout << x << " ";	
	}	

	return 0;
}
